cogent/src/Cogent/DataLayout/CodeGen.hs

```diff
@@ -0,0 +1,133 @@
+  functionNameIdentifier
+  =
+  ( CFnDefn
+  
+    -- (return type, function name)
+    ( unsignedIntType, functionNameIdentifier )
+    
+    -- [(param type, param name)]
+    [ ( CPtr unsignedIntType, ptrIdentifier ) ]
+    
+    -- statements
+    [ CBIStmt
+      ( CReturn
+        ( Just 
+          ( CBinOp And
+            ( CBinOp Rsh
+              ( CDeref ( CBinOp Add ptrVariable wordOffsetLiteral ) )
+              bitOffsetLiteral
+            )
+            maskLiteral
+          )
+        )
+      )
+    ]
+  
+    staticInlineFnSpec
+  )
+  where
+    ptrIdentifier = "p"
+    ptrVariable   = CVar ptrIdentifier Nothing
+    
+    wordOffsetLiteral = unsignedIntLiteral wordOffsetABR
+    bitOffsetLiteral  = unsignedIntLiteral bitOffsetABR
+    maskLiteral   = unsignedIntLiteral $ sizeToMask bitSizeABR
+
+
+{- | Creates a C function to extract set the contents of an
+     AlignedBitRange in a Cogent boxed type.
+     
+     `alignedRangeSetter AlignedBitRange { bitSizeABR, bitOffsetABR, wordOffsetABR} functionNameIdentifier` should be the C function
+     
+     @
+      static inline void set`functionNameIdentifier`(unsigned int *p, unsigned int v) {
+        *(p + `wordOffsetABR`)
+          = *(p + `wordOffsetABR`)
+
+          // clear the bits
+          & ~(`sizeToMask bitSizeABR` << `bitOffsetABR`)) 
+          
+          // set the bits
+          | ((`sizeToMask bitSizeABR` & v) << `bitOffsetABR`);
+      }
+     @
+-}
+genAlignedRangeSetter :: AlignedBitRange -> CId -> CExtDecl
+genAlignedRangeSetter
+  AlignedBitRange { bitSizeABR, bitOffsetABR, wordOffsetABR }
+  functionNameIdentifier
+  =
+  ( CFnDefn
+  
+    -- (return type, function name)
+    ( CVoid, functionNameIdentifier )
+    
+    -- [(param type, param name)]
+    [ ( CPtr unsignedIntType, ptrIdentifier )
+    , ( unsignedIntType, valueIdentifier )
+    ]
+    
+    -- statements
+    [ CBIStmt
+      ( CAssign
+        ( CDeref ( CBinOp Add ptrVariable wordOffsetLiteral ) )
+        ( CBinOp Or
+          ( CBinOp And
+            ( CDeref ( CBinOp Add ptrVariable wordOffsetLiteral ) )
+            ( CUnOp Not ( CBinOp Lsh maskLiteral bitOffsetLiteral ) )
+          )
+          ( CBinOp Lsh
+            ( CBinOp And maskLiteral valueVariable )
+            bitOffsetLiteral
+          )
+        )
+      )
+    ]
+    
+    staticInlineFnSpec
+  )
+  where
+    ptrIdentifier = "p"
+    ptrVariable   = CVar ptrIdentifier Nothing
+    
+    valueIdentifier = "v"
+    valueVariable   = CVar valueIdentifier Nothing
+    
+    wordOffsetLiteral = unsignedIntLiteral wordOffsetABR
+    bitOffsetLiteral  = unsignedIntLiteral bitOffsetABR
+    maskLiteral   = unsignedIntLiteral $ sizeToMask bitSizeABR
+    
+    
+sizeToMask n
+  | 0 <= n && n <= 8  = 2^n - 1
+  | otherwise         = __impossible $ "After alignment"
+    
+-- Produces syntax for an unsigned integer literal with the given value
+unsignedIntLiteral :: Integer -> CExpr
```

Comment:
I have similar functions in `Cogent.C.Compile`.

Reply:
I can't find any relevant ones, but for the moment am trying to minimise depending on `Cogent.C.Compile` because it will have to depend on `Cogent.DataLayout.CodeGen`, and thus will need to remove `Cogent.DataLayout.CodeGen`s dependency on `Cogent.C.Compile`.

Reply:
Ok, point taken. The code is here: https://github.com/NICTA/cogent/blob/master/cogent/src/Cogent/C/Compile.hs#L159